### PR TITLE
Adapting process to run off GitHub actions

### PR DESCRIPTION
**Pull request by edmoss345**

### CODE REVIEW

File: __pycache__/srh_models.cpython-311.pyc
Comment (istride) [inline __pycache__/srh_models.cpython-311.pyc:1]: pyc files are compiled and don't need to be committed. It's best to add them to the .gitignore file.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Repo for SRH_Jamaica
+
+## Setup
+1. Install python `>=3.6`
+2. Run `pip install -r requirements.txt`
+3. Run `npm install`
```
Comment (istride) [inline README.md]: Node/npm must be installed first, and for that we should recommend a version. My recommendation is latest LTS (Long-term support) version.

File: models/srh_models.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+class SgEntryModel(DataRowModel):
+	question: str = ''
+	yes_flow: str = ''
+	no_message: str = ''
+
+
+
+
+
+
```
Comment (istride) [inline models/srh_models.py]: Clean up this whitespace

File: .gitignore
Comment (istride) [inline .gitignore:1]: I think we should add the output directory to gitignore too

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  workflow_dispatch:
+
+jobs:
+
+  core-function:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
```
Comment (istride) [inline .github/workflows/main.yml]: Make sure we are using the latest versions of these actions. I think checkout is at v3 at the moment.

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+      - name: Install node dependencies
+        run: npm ci
+
+      - name: Install python dependenceis
```
Comment (istride) [inline .github/workflows/main.yml]: Spelling.

File: srh_jamaica.ps1
Comment (istride) [inline srh_jamaica.ps1:1]: Is this file still necessary now that has an equivalent in Python? I would like there to be one source of truth for how to do the processing.

File: srh_jamaica.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+import os
+import subprocess
+import sys
+from rapidpro_flow_tools import flow_converter
+
+def main(credentials = None, token = None):
+
+    credentials = os.getenv('CREDENTIALS', credentials)
+    token = os.getenv('TOKEN', token)
```
Comment (istride) [inline srh_jamaica.py]: Move these out of `main` and pass them in as arguments.

File: srh_jamaica.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        # JSON_FILENAME = "..\\srh-jamaica-chatbot\\flows\\" + source_file_name + ".json"
+        # source_file_name = source_file_name + "_avatar"
+        # CONFIG_ab = "..\\srh-jamaica-chatbot\\edits\\ab_config.json"
+        # output_path_2 = "temp\\" + source_file_name + ".json"
+        # AB_log = "..\\srh-jamaica-chatbot\\temp\\AB_warnings.log"
+        # os.chdir("..\\rapidpro_abtesting")
+        # subprocess.run(["python", "main.py", JSON_FILENAME, "..\\srh-jamaica-chatbot\\" + output_path_2, SPREADSHEET_ID, "--format", "google_sheets", "--logfile", AB_log, "--config=" + CONFIG_ab])
+        # print("added A/B tests and localisation")
+        # input_path = output_path_2
+        # os.chdir("..\\srh-jamaica-chatbot")
```
Comment (istride) [inline srh_jamaica.py:40]: No commented code, please. It's either necessary or it's removed. If it's only required sometimes, find a way to toggle it on/off at runtime.

**Review by istride — CHANGES_REQUESTED** [review]
Needs a bit of cleaning up.

File: .github/workflows/main.yml (hunk 4 above)
Comment (edmoss345) [reply .github/workflows/main.yml]: changed

File: .github/workflows/main.yml (hunk 3 above)
Comment (edmoss345) [reply .github/workflows/main.yml]: changed

File: .gitignore
Comment (edmoss345) [reply .gitignore:1]: have added

File: README.md (hunk 1 above)
Comment (edmoss345) [reply README.md]: have added in as a step

File: __pycache__/srh_models.cpython-311.pyc
Comment (edmoss345) [reply __pycache__/srh_models.cpython-311.pyc:1]: have added to gitignore

File: models/srh_models.py (hunk 2 above)
Comment (edmoss345) [reply models/srh_models.py]: have removed

File: srh_jamaica.ps1
Comment (edmoss345) [reply srh_jamaica.ps1:1]: will look to remove in due course but will double check that Chiara is happy running the python file before I take this out

File: srh_jamaica.py (hunk 5 above)
Comment (edmoss345) [reply srh_jamaica.py]: have moved to arguments in actions file as suggested

File: srh_jamaica.ps1
Comment (edmoss345) [reply srh_jamaica.ps1:1]: @fagiothree just tagging you here to remind me as a point we need to discuss, need to check that the python file is usable for you, if so we can remove the powershell

File: srh_jamaica.py (hunk 6 above)
Comment (edmoss345) [reply srh_jamaica.py:40]: @fagiothree just tagging you here to remind me as a point we need to discuss, when we do/don't need this code and the best way to handle it

File: srh_jamaica.py (hunk 5 above)
Comment (istride) [reply srh_jamaica.py]: Sorry, I should have explained more clearly. I think it's fine to get `credentials` and `token` from environment variables, my issue was with how it was implemented. I think it should be something like the following:

```python
if __name__ == '__main__':
    credentials = os.getenv('CREDENTIALS')
    token = os.getenv('TOKEN')
    main(credentials, token)
```

We read the environment variables only if this module is called as a script (i.e. `python srh_jamaica.py`), then pass them over to the `main` function.

File: srh_jamaica.py (hunk 6 above)
Comment (istride) [reply srh_jamaica.py:40]: After discussion, it has been concluded that this code is likely to be required in the future, so we need to keep it. Leave it as is and I'll create a separate issue to deal with it later.

File: .gitignore [hunk 7]
```diff
@@ -121,3 +124,6 @@ dist
 .yarn/build-state.yml
 .yarn/install-state.gz
 .pnp.*
+
+# Comiled pyc files
+*.pyc
```
Comment (istride) [inline .gitignore:129]: This is fine, but you still need to delete all the '\_\_pycache\_\_' directories that have already been committed.

File: requirements.txt [hunk 8]
```diff
@@ -0,0 +1,8 @@
+Jinja2~=3.0.3
+networkx~=2.5.1
+pydantic~=1.8.2
+tablib[ods]>=3.1.0
+openpyxl~=3.0.7
+google-api-python-client~=2.6.0
+google-auth-oauthlib~=0.4.4
+rapidpro-flow-tools
```
Comment (istride) [inline requirements.txt]: I think we should constrain `rapidpro-flow-tools` to specific versions, like everything else.

File: srh_jamaica.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
+import os
+import subprocess
+import sys
+from rapidpro_flow_tools import flow_converter
+
+def main(credentials = None, token = None):
+
+    source_files = ["srh_registration", "srh_entry", "srh_content", "srh_safeguarding"]
+    spreadsheet_IDS = ["1yett-Rfzb9Ou8IQ1kwtrKPN_auhM-lk66r9gkqNV1As", "19xvYfwWKA1hT5filGPWYEobQL1ZFfcFbTj1-aJCN8OQ", "1hOlgdqjmXZgl51L1olt357Gfiw2zRHNEl98aYTf8Hwo", "1A_p3cb3KNgX8XxD9MlCIoa294Y4Pb9obUKfwIvERALY"]
```
Comment (istride) [inline srh_jamaica.py]: These two lists should be combined, so that it's clear what the relationship between them is.

```python
sources = [
    {"filename": "srh_registration", "spreadsheet_id": "1yett-Rfzb9Ou8IQ1kwtrKPN_auhM-lk66r9gkqNV1As"},
    // etc...
]

// usage
for source in sources:
    source_file_name = source["filename"]
    spreadsheet_id  = source["spreadsheet_id"]
```

Alternatively, a class could be created.

```python
class Source:
    def __init__(self, filename, spreadsheet_id):
        self.filename = filename
        self.spreadsheet_id = spreadsheet_id

sources = [
    Source("srh_registration", "1yett-Rfzb9Ou8IQ1kwtrKPN_auhM-lk66r9gkqNV1A"),
    // etc...
]

for source in sources:
    source_file_name = source.filename
    sid = source.spreadsheet_id
```

File: .gitignore [hunk 10]
```diff
@@ -1,7 +1,10 @@
 # Intermediary output files
 temp/
+output/
```
Comment (istride) [inline .gitignore:3]: Also need to delete the 'output' directory that has already been committed.

**Review by istride — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+      - name: Install python dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r requirements.txt  
+
+      - name: Run Python function
+        env:
+          CREDENTIALS: ${{ secrets.CREDENTIALS }}
+          TOKEN: ${{ secrets.TOKEN }}
```
Comment (istride) [inline .github/workflows/main.yml:25]: I think only credentials will be required. You get a token by authenticating yourself with the credentials, then use the token to authorise you to use other Google APIs.

File: .gitignore (hunk 7 above)
Comment (edmoss345) [reply .gitignore:129]: have deleted

File: requirements.txt (hunk 8 above)
Comment (edmoss345) [reply requirements.txt]: have specified a version as suggested

File: .github/workflows/main.yml (hunk 11 above)
Comment (istride) [reply .github/workflows/main.yml:25]: We can work this out after merging.

File: models/__pycache__/srh_models.cpython-311.pyc
Comment (istride) [inline models/__pycache__/srh_models.cpython-311.pyc:1]: Remove

**Review by istride — CHANGES_REQUESTED** [review]
It's just occurred to me that the 'flows' directory contains the Google sheets converted to JSON. It doesn't make sense to store them in this repo, but it would probably be useful to add them to the output artifact i.e. "Publish output folder" step. It would probably help with debugging, if nothing else.

File: srh_jamaica.py (hunk 9 above)
Comment (edmoss345) [reply srh_jamaica.py]: have combined as suggested

File: srh_jamaica.py (hunk 5 above)
Comment (edmoss345) [reply srh_jamaica.py]: have moved as suggested

File: models/__pycache__/srh_models.cpython-311.pyc
Comment (edmoss345) [reply models/__pycache__/srh_models.cpython-311.pyc:1]: have removed

Comment (edmoss345) [reply models/__pycache__/srh_models.cpython-311.pyc:1]: have changed everything to go to the output as suggested

**Comment by edmoss345** [comment]
> It's just occurred to me that the 'flows' directory contains the Google sheets converted to JSON. It doesn't make sense to store them in this repo, but it would probably be useful to add them to the output artifact i.e. "Publish output folder" step. It would probably help with debugging, if nothing else.

Good point, have now directed everything to the outputs folder

File: srh_jamaica.ps1
Comment (edmoss345) [reply srh_jamaica.ps1:1]: think we should leave the powershell file for now as we have made semi-significant changes, just in case Chiara needs to re-run something quickly with a script she is familiar with

**Review by istride — APPROVED** [review]
